File: integrations/langgraph/python/examples/agents/a2ui_dynamic_schema/agent.py
"""
Dynamic A2UI tool: LLM-generated UI from conversation context.

A secondary LLM generates v0.9 A2UI components via a structured tool call.
The generate_a2ui tool wraps the output as a2ui_operations, which the
middleware detects in the TOOL_CALL_RESULT and renders automatically.
"""

import json
import os
from typing import Any, List

from langchain.tools import tool, ToolRuntime
from langchain_core.messages import SystemMessage
from langchain_core.tools import tool as lc_tool
from langchain_core.runnables import RunnableConfig
from langchain_openai import ChatOpenAI
from langgraph.graph import StateGraph, END, MessagesState
from langgraph.prebuilt import ToolNode
from ag_ui_langgraph import get_a2ui_tools


CUSTOM_CATALOG_ID = "https://a2ui.org/demos/dojo/dynamic_catalog.json"

# Project-specific composition rules — tells the subagent how to use the
# pre-made domain components (HotelCard, ProductCard, TeamMemberCard) shipped
# in the dojo's dynamic catalog.
COMPOSITION_GUIDE = """
## Available Pre-made Components

You have 4 components. Use Row as the root with structural children to repeat a card per item.

### Row
Layout container. Use structural children to repeat a card template:
  {"id":"root","component":"Row","children":{"componentId":"card","path":"/items"}}

### HotelCard
Props: name, location, rating (number 0-5), pricePerNight, amenities (optional), action
Example:
  {"id":"card","component":"HotelCard","name":{"path":"name"},"location":{"path":"location"},
   "rating":{"path":"rating"},"pricePerNight":{"path":"pricePerNight"},
   "action":{"event":{"name":"book","context":{"name":{"path":"name"}}}}}

### ProductCard
Props: name, price, rating (number 0-5), description (optional), badge (optional), action
Example:
  {"id":"card","component":"ProductCard","name":{"path":"name"},"price":{"path":"price"},
   "rating":{"path":"rating"},"description":{"path":"description"},
   "action":{"event":{"name":"select","context":{"name":{"path":"name"}}}}}

### TeamMemberCard
Props: name, role, department (optional), email (optional), avatarUrl (optional), action
Example:
  {"id":"card","component":"TeamMemberCard","name":{"path":"name"},"role":{"path":"role"},
   "department":{"path":"department"},"email":{"path":"email"},
   "action":{"event":{"name":"contact","context":{"name":{"path":"name"}}}}}

## RULES
- Root is ALWAYS a Row with structural children: {"componentId":"<card-id>","path":"/items"}
- Inside templates, use RELATIVE paths (no leading slash): {"path":"name"} not {"path":"/name"}
- Always provide data in the "data" argument as {"items":[...]}
- Pick the card type that best matches the user's request
- Generate 3-4 realistic items with diverse data
"""

base_model = ChatOpenAI(model="gpt-4o")

TOOLS = [
    get_a2ui_tools(
        model=base_model,
        default_catalog_id=CUSTOM_CATALOG_ID,
        composition_guide=COMPOSITION_GUIDE,
    )
]


class AgentState(MessagesState):
    tools: List[Any]
    copilotkit: dict  # CopilotKit context (actions, etc.)

# LangGraph requires state keys declared in the schema.
# "ag-ui" uses a hyphen which isn't valid as a Python identifier,
# so we patch it into the annotations directly.
AgentState.__annotations__["ag-ui"] = dict


SYSTEM_PROMPT = """You are a helpful assistant that creates rich visual UI on the fly.

When the user asks for visual content (product comparisons, dashboards, lists, cards, etc.),
use the generate_a2ui tool to create a dynamic A2UI surface.
IMPORTANT: After calling the tool, do NOT repeat the data in your text response. The tool renders UI automatically. Just confirm what was rendered."""


async def chat_node(state: AgentState, config: RunnableConfig):
    model = base_model.bind_tools(TOOLS, parallel_tool_calls=False)

    response = await model.ainvoke([
        SystemMessage(content=SYSTEM_PROMPT),
        *state["messages"],
    ], config)

    return {"messages": [response]}


def route_after_chat(state: AgentState):
    last_message = state["messages"][-1]
    if hasattr(last_message, "tool_calls") and last_message.tool_calls:
        return "tool_node"
    return END


workflow = StateGraph(AgentState)
workflow.add_node("chat_node", chat_node)
workflow.add_node("tool_node", ToolNode(tools=TOOLS))
workflow.set_entry_point("chat_node")
workflow.add_conditional_edges("chat_node", route_after_chat)
workflow.add_edge("tool_node", "chat_node")

is_fast_api = os.environ.get("LANGGRAPH_FAST_API", "false").lower() == "true"

if is_fast_api:
    from langgraph.checkpoint.memory import MemorySaver
    memory = MemorySaver()
    graph = workflow.compile(checkpointer=memory)
else:
    graph = workflow.compile()

File: integrations/langgraph/typescript/examples/src/agents/a2ui_fixed_schema/agent.ts
/**
 * Fixed-schema A2UI agent (prebuilt).
 *
 * Pre-built component layouts for flight and hotel cards. The agent only
 * supplies the data; layout/styling is fixed in code. Demonstrates the
 * "controlled gen-UI" pattern: author owns the UI shape, agent owns the data.
 */

import { createAgent } from "langchain";
import { MemorySaver } from "@langchain/langgraph";
import { copilotkitMiddleware } from "@copilotkit/sdk-js/langgraph";
import { tool } from "@langchain/core/tools";

const CUSTOM_CATALOG_ID =
  "https://a2ui.org/demos/dojo/fixed_catalog.json";

const A2UI_OPERATIONS_KEY = "a2ui_operations";

// Flight search layout — agent supplies `flights` array; rendering is fixed.
const FLIGHT_SURFACE_ID = "flight-search-results";
const FLIGHT_SCHEMA: Array<Record<string, unknown>> = [
  {
    id: "root",
    component: "Row",
    children: { componentId: "flight-card", path: "/flights" },
    gap: 16,
  },
  {
    id: "flight-card",
    component: "FlightCard",
    airline: { path: "airline" },
    airlineLogo: { path: "airlineLogo" },
    flightNumber: { path: "flightNumber" },
    origin: { path: "origin" },
    destination: { path: "destination" },
    date: { path: "date" },
    departureTime: { path: "departureTime" },
    arrivalTime: { path: "arrivalTime" },
    duration: { path: "duration" },
    status: { path: "status" },
    price: { path: "price" },
    action: {
      event: {
        name: "book_flight",
        context: {
          flightNumber: { path: "flightNumber" },
          origin: { path: "origin" },
          destination: { path: "destination" },
          price: { path: "price" },
        },
      },
    },
  },
];

// Hotel search layout — agent supplies `hotels` array; rendering is fixed.
const HOTEL_SURFACE_ID = "hotel-search-results";
const HOTEL_SCHEMA: Array<Record<string, unknown>> = [
  {
    id: "root",
    component: "Row",
    children: { componentId: "hotel-card", path: "/hotels" },
    gap: 16,
  },
  {
    id: "hotel-card",
    component: "HotelCard",
    name: { path: "name" },
    location: { path: "location" },
    rating: { path: "rating" },
    pricePerNight: { path: "price" },
    action: {
      event: {
        name: "book_hotel",
        context: {
          hotelName: { path: "name" },
          price: { path: "price" },
        },
      },
    },
  },
];

function renderOperations(
  surfaceId: string,
  catalogId: string,
  schema: Array<Record<string, unknown>>,
  data: Record<string, unknown>,
): string {
  const ops = [
    {
      version: "v0.9",
      createSurface: { surfaceId, catalogId },
    },
    {
      version: "v0.9",
      updateComponents: { surfaceId, components: schema },
    },
    {
      version: "v0.9",
      updateDataModel: { surfaceId, path: "/", value: data },
    },
  ];
  return JSON.stringify({ [A2UI_OPERATIONS_KEY]: ops });
}

const searchFlights = tool(
  async ({ flights }: { flights: Array<Record<string, unknown>> }) => {
    return renderOperations(
      FLIGHT_SURFACE_ID,
      CUSTOM_CATALOG_ID,
      FLIGHT_SCHEMA,
      { flights },
    );
  },
  {
    name: "search_flights",
    description:
      "Search for flights and display the results as rich cards. Each flight " +
      "must have: id, airline (e.g. 'United Airlines'), airlineLogo (use Google " +
      "favicon API like 'https://www.google.com/s2/favicons?domain=united.com&sz=128'), " +
      "flightNumber, origin, destination, date (e.g. 'Tue, Mar 18'), departureTime, " +
      "arrivalTime, duration (e.g. '4h 25m'), status ('On Time' or 'Delayed'), " +
      "and price (e.g. '$289').",
    schema: {
      type: "object",
      properties: {
        flights: {
          type: "array",
          items: { type: "object" },
          description: "Array of flight result objects.",
        },
      },
      required: ["flights"],
    } as any,
  },
);

const searchHotels = tool(
  async ({ hotels }: { hotels: Array<Record<string, unknown>> }) => {
    return renderOperations(
      HOTEL_SURFACE_ID,
      CUSTOM_CATALOG_ID,
      HOTEL_SCHEMA,
      { hotels },
    );
  },
  {
    name: "search_hotels",
    description:
      "Search for hotels and display the results as rich cards with star ratings. " +
      "Each hotel must have: id, name (e.g. 'The Plaza'), location " +
      "(e.g. 'Midtown Manhattan, NYC'), rating (float 0-5, e.g. 4.5), and " +
      "price (per night, e.g. '$350'). Generate 3-4 realistic results.",
    schema: {
      type: "object",
      properties: {
        hotels: {
          type: "array",
          items: { type: "object" },
          description: "Array of hotel result objects.",
        },
      },
      required: ["hotels"],
    } as any,
  },
);

const checkpointer = new MemorySaver();

export const a2uiFixedSchemaGraph = createAgent({
  model: "openai:gpt-4o",
  tools: [searchFlights, searchHotels],
  middleware: [copilotkitMiddleware],
  systemPrompt: `You are a helpful travel assistant that can search for flights and hotels.

When the user asks about flights, use the search_flights tool.
When the user asks about hotels, use the search_hotels tool.
IMPORTANT: After calling a tool, do NOT repeat or summarize the data in your text response. The tool renders a rich UI automatically. Just say something brief like "Here are your results" or ask if they'd like to book.

For flights, each needs: id, airline, airlineLogo (Google favicon API), flightNumber, origin, destination,
date, departureTime, arrivalTime, duration, status, statusIcon, and price.

For hotels, each needs: id, name, location, rating (float 0-5), and price (per night).

Generate 3-5 realistic results.`,
  checkpointer,
});
